src/agent/thread_helper.cpp

```diff
@@ -0,0 +1,361 @@
+    }
+
+    return extPanId;
+}
+
+void ThreadHelper::Attach(const std::string &         aNetworkName,
+                          uint16_t                    aPanId,
+                          uint64_t                    aExtPanId,
+                          const std::vector<uint8_t> &aMasterKey,
+                          const std::vector<uint8_t> &aPSKc,
+                          uint32_t                    aChannelMask,
+                          ResultHandler               aHandler)
+
+{
+    otError         err = OT_ERROR_NONE;
+    otExtendedPanId extPanId;
+    otMasterKey     masterKey;
+    otPskc          pskc;
+    otError         pskcRet;
+    uint32_t        channelMask;
+    uint8_t         channel;
+
+    VerifyOrExit(aHandler != nullptr, err = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(mAttachHandler == nullptr && mJoinerHandler == nullptr, err = OT_ERROR_INVALID_STATE);
+    mAttachHandler = aHandler;
+    VerifyOrExit(aMasterKey.empty() || aMasterKey.size() == sizeof(masterKey.m8), err = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aPSKc.empty() || aPSKc.size() == sizeof(pskc.m8), err = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aChannelMask != 0, err = OT_ERROR_INVALID_ARGS);
+
+    if (aPanId == UINT16_MAX)
+    {
+        RandomFill(&aPanId, sizeof(aPanId));
+    }
+
+    if (aExtPanId != UINT64_MAX)
+    {
+        extPanId = ToOtExtendedPanId(aExtPanId);
+    }
+    else
+    {
+        RandomFill(extPanId.m8, sizeof(extPanId.m8));
+    }
+
+    if (!aMasterKey.empty())
+    {
+        memcpy(masterKey.m8, &aMasterKey[0], sizeof(masterKey.m8));
+    }
+    else
+    {
+        RandomFill(masterKey.m8, sizeof(masterKey.m8));
+    }
+
+    if (!aPSKc.empty())
+    {
+        memcpy(pskc.m8, &aPSKc[0], sizeof(pskc.m8));
+    }
+    else
+    {
+        RandomFill(pskc.m8, sizeof(pskc.m8));
+    }
+
+    SuccessOrExit(err = otThreadSetNetworkName(mInstance, aNetworkName.c_str()));
+    SuccessOrExit(err = otLinkSetPanId(mInstance, aPanId));
+    SuccessOrExit(err = otThreadSetExtendedPanId(mInstance, &extPanId));
+    SuccessOrExit(err = otThreadSetMasterKey(mInstance, &masterKey));
+
+    channelMask = otPlatRadioGetPreferredChannelMask(mInstance) & aChannelMask;
+
+    if (channelMask == 0)
+    {
+        channelMask = otLinkGetSupportedChannelMask(mInstance) & aChannelMask;
+    }
+    VerifyOrExit(channelMask != 0, err = OT_ERROR_INVALID_ARGS);
+
+    channel = RandomChannelFromChannelMask(channelMask);
+    SuccessOrExit(otLinkSetChannel(mInstance, channel));
+
+    pskcRet = otThreadSetPskc(mInstance, &pskc);
+    VerifyOrExit(pskcRet == OT_ERROR_NONE, err = OT_ERROR_FAILED);
+
+    if (!otIp6IsEnabled(mInstance))
+    {
+        SuccessOrExit(err = otIp6SetEnabled(mInstance, true));
+    }
+    SuccessOrExit(err = otThreadSetEnabled(mInstance, true));
+exit:
+    if (err != OT_ERROR_NONE)
+    {
+        if (aHandler)
+        {
+            aHandler(err);
+        }
+        mAttachHandler = nullptr;
+    }
```

Comment:
No, what's the behavior of attaching in wpantund, does it have a hard-coded timeout value since there's not a timeout argument in the attach dbus call?

Reply:
Another concern is that for a SED device, it may stay in detached state for a long time until a valid parent comes. Since the Attach calls are triggered in a async manner, I think it acceptable to not have a timeout here.

Reply:
Do we not need the `mAttachHandler` to prepare and emit the dbus response?

Reply:
We always need the mAttachHandler to be called to send the dbus response.  When the SED device waiting for a parent, we shall not send the response however.